### PR TITLE
Add type hints to media rest resources.

### PR DESCRIPTION
**Pull request by clokep**
I was tracking down a couple of sentry issues in the media code and figured it'd be helpful to add types throughout. This should be reviewable commit by commit.

### CODE REVIEW

**Comment by clokep** [comment]
Actually I need to look at this a bit closer. I think one of the changes is incorrect.

File: synapse/rest/media/v1/_base.py [hunk 1]
```diff
@@ -153,7 +163,8 @@ def _quote(x):
     # select private. don't bother setting Expires as all our
     # clients are smart enough to be happy with Cache-Control
     request.setHeader(b"Cache-Control", b"public,max-age=86400,s-maxage=86400")
-    request.setHeader(b"Content-Length", b"%d" % (file_size,))
+    if file_size is not None:
+        request.setHeader(b"Content-Length", b"%d" % (file_size,))
```
Comment (clokep) [inline synapse/rest/media/v1/_base.py:167]: This is due to `respond_with_responder` accept `None` which then gets passed through to here. I don't know if that *actually* gets called with `None` ever though. 😢

File: synapse/rest/media/v1/filepath.py [hunk 2]
```diff
@@ -16,11 +17,12 @@
 import functools
 import os
 import re
+from typing import Callable, List
 
 NEW_FORMAT_ID_RE = re.compile(r"^\d\d\d\d-\d\d-\d\d")
 
 
-def _wrap_in_base_path(func):
+def _wrap_in_base_path(func: "Callable[..., str]") -> "Callable[..., str]":
```
Comment (erikjohnston) [inline synapse/rest/media/v1/filepath.py:25]: Why the quotes?

**Review by erikjohnston — APPROVED** [review]

File: synapse/rest/media/v1/filepath.py (hunk 2 above)
Comment (clokep) [reply synapse/rest/media/v1/filepath.py:25]: I believe that `...` breaks on Python 3.5? We seem to quote similar things in synapse/storage/database.py, but not other places...

Comment (clokep) [reply synapse/rest/media/v1/filepath.py:25]: https://github.com/matrix-org/synapse/blob/1315a2e8be702a513d49c1142e9e52b642286635/synapse/storage/database.py#L177-L178

**Pull request merged**